### ISSUE TITLE
Compressed NNUE weights size

### ISSUE THREAD

**Issue by ddobbelaere** [official-stockfish/Stockfish#3274]
I just did a small experiment (namely "`xz -z -9e`" (max. xz compression)) on some NNUE nets of https://tests.stockfishchess.org/nns
Here are the resulting file sizes in bytes:

Sergiovieri run and derivatives (leading to all recent master nets):
```
9776088 nn-82215d0fd0df.nnue.xz (sergiovieri) <-- master^10
9721312 nn-112bb1c8cdb5.nnue.xz (sergiovieri) <-- master^11
9642504 nn-62ef826d1a6d.nnue.xz (SFisGOD)     <-- master
9642504 nn-c3ca321c51c9.nnue.xz (SFisGOD)     <-- master^1
9642504 nn-cb26f10b1fd9.nnue.xz (Fisherman)   <-- master^2
9642504 nn-2eb2e0707c2b.nnue.xz (SFisGOD)     <-- master^3
9642504 nn-eba324f53044.nnue.xz (vdv)         <-- master^4
9642504 nn-04cf2b4ed1da.nnue.xz (vdv)         <-- master^5
9642504 nn-baeb9ef2d183.nnue.xz (SFisGOD)     <-- master^6
9642504 nn-04a843f8932e.nnue.xz (vdv)         <-- master^7
9642504 nn-03744f8d56d8.nnue.xz (sergiovieri) <-- master^8
9642500 nn-6554800d4b0c.nnue.xz (Fisherman)
9639372 nn-308d71810dff.nnue.xz (sergiovieri) <-- master^9
```

Other runs:
```
8772772 nn-30f6ddc579c2.nnue.xz (noobpwnftw)
8771688 nn-91fec6f6756a.nnue.xz (noobpwnftw)
8613628 nn-ddbf15bd12bd.nnue.xz (vdv)
8589724 nn-64fc1e0029b5.nnue.xz (noobpwnftw)
8589528 nn-31c47f9dbd91.nnue.xz (noobpwnftw)
7680524 nn-be652a0c885a.nnue.xz (vdv)
6928364 nn-01de8d92ba3e.nnue.xz (glinscott)
6775392 nn-11c4729f1ece.nnue.xz (vdv)
5923328 nn-26b7fcdb21c6.nnue.xz (vdv)
2563132 nn-a20699671e55.nnue.xz (MiauiKatze)
```

It seems that the compressed weights of the successful run of @sergiovieri, along with its derivatives (mostly post-optimizations of final layer) are at least about 1MB bigger than all other independent runs.

If this size is a measure of entropy (information content), it seems these networks encode more knowledge. Maybe regularization is too strong for the other independent runs (this is just a bold hypothesis), dragging the weights more towards zero and forgetting previously learned knowledge more quickly?

**Comment by ddobbelaere** [comment]
Apparently the pytorch trainer doesn't explicitly use an L2 regularization term as part of the loss function (as opposed to AlphaZero/Leela Chess Zero).

An optimizer called Ranger is used that does al the grad-based training.

One thing I read from the docs is: "Best training results - use a 75% flat lr, then step down and run lower lr for 25%, or cosine descend last 25%."

It's not obvious to me that the latter part is supported by the current pytorch trainer, or am I mistaken (I see only a single flat LR)?

**Comment by vondele** [comment]
right now it is a flat LR. I think it could be worthwhile, but nobody wrote the code for it yet..

**Comment by ddobbelaere** [comment]
Hmm, ok. My limited experience from (following) Leela Chess Zero, where the LR is piecewise constant during training (typically factor 10 smaller at each so-called LR drop, if loss has converged), is that the final nets are typically hundreds of elo stronger than best nets at first LR region. After each LR drop (of which there are typically three), the loss function typically converges in exponential fashion to a lower plateau (on your loss curves I only see one such regime).

I have no experience with Ranger, but anyway they recommend a final "annealing" phase that gradually drops LR, so this seems very promising to me.

**Comment by ddobbelaere** [comment]
I just hacked together a script based on https://github.com/glinscott/nnue-pytorch/blob/master/serialize.py and https://hxim.github.io/Stockfish-Evaluation-Guide/ to investigate the weights of the input to feature transformer.

Let's load all the weights w and visualize min(128, abs(w)) for the current master net (nn-62ef826d1a6d.nnue) and some the best vdv net tested on fishtest (nn-ddbf15bd12bd.nnue).

The results are surprising:

![image](https://user-images.githubusercontent.com/38748663/103217067-fc3f1480-4917-11eb-8bdf-1f5d19dad1c5.png)
![image](https://user-images.githubusercontent.com/38748663/103217099-0cef8a80-4918-11eb-82a3-22afafb28464.png)

A red box of size 304x128 corresponds to all the weights connected to one feature (there are 256 boxes).

Note that in the vdv net there are features that are as good as dead (not excited at all!). Even stranger, not all weights are exactly, but some are also +/-2, but never +/-1!

I hope there is not a big issue with training, e.g. dying ReLu issue in final layers (see https://en.wikipedia.org/wiki/Rectifier_(neural_networks)#Potential_problems), which can be caused by absence of regularization.

@vondele If you have any other (better?) nets to investigate, let me know.

**Comment by ddobbelaere** [comment]
Same exercise for best noob net on fishtest (nn-64fc1e0029b5.nnue). No visual indication related to dead features here at first glance.
 @noobpwnftw Did you use Nodchip trainer for this net or pytorch trainer?

![image](https://user-images.githubusercontent.com/38748663/103219518-3a3f3700-491e-11eb-9ee9-a3eb02ad635e.png)

**Comment by ddobbelaere** [comment]
Now that I think of it, a final "normalization" before exporting to quantized nnue format might be possible (and highly desirable!) as follows:

**Ensure that dynamic range of weights for each input feature is high enough by scaling biases/weights in layers that follow, such that the non-quantized net is equivalent to the initial net.**

**Comment by vondele** [comment]
@ddobbelaere interesting analysis. It indeed looks like this could be an issue with the training.

noob is using the nodchip trainer.

**Comment by vondele** [comment]
BTW, do you have a link to your script? Even better, what about a PR to the pytorch trainer.

**Comment by Sopel97** [comment]
I find the results kinda weird. We observed dying relu problem with the nodchip trainer but it only happened with very high LR and after the first initial batch. The pytorch trainer uses a much lower LR (and requires much longer training to get a good net). They both use the same loss function, nodchip defaults to LR=1 which is fine, ranger has LR=1e-3 (unless it does something weird at the start?). So this is likely not cause by the learning rate.

**Comment by ddobbelaere** [comment]
The script can be found at https://gist.github.com/ddobbelaere/ad4e2645828b3fbd771249e37cfc2d0a. I've warned you it's hacky :). Unfortunately, I don't have the time at the moment to create a PR for pytorch trainer with a proper "visualizer".

**Comment by ddobbelaere** [comment]
@Sopel97 Yeah, I also don't know what the issue is to be honest (if there is any).

I just did a bit more research on the feature transformer layer. I'm pretty sure the clipped ReLUs of the "dead features" of the vdv nets are not dying at least, as can be attested by their small (negative) bias terms:

![image](https://user-images.githubusercontent.com/38748663/103346140-7ea01380-4a93-11eb-922a-73220f80d7f3.png)

Maybe the combination of training and input data had no way to extract any more useful features back then?

I also did a brief investigation on the first FC layer, but see no obvious weird things...

**Comment by ddobbelaere** [comment]
Same observations on latest vdv uploaded to fishtest (nn-535ee551b2cf.nnue). Some internal features seem dead/unused:

![image](https://user-images.githubusercontent.com/38748663/103350706-27a13b00-4aa1-11eb-92d2-96e1f7fec7b8.png)

**Comment by Sopel97** [comment]
> Or maybe the combination of training and input data had no way to extract any more useful features back then? Or maybe a bad weights initialization?

Could actually be initialization. Nodchip initializes the bias to 0.5 +- random. pytorch does 0 +- random.

**Comment by ddobbelaere** [comment]
Hmm, yeah that could explain things. A large negative bias is a potential kill shot for the ReLU.

One of the things that is probably uncommon in our architecture (w.r.t. other nets for e.g. image recognition) is that our excitation is _sparse_ (limited number of excited input features), which means that there is potentially not enough "input energy" to overcome a large negative bias. If I am not mistaken (see e.g. https://stackoverflow.com/questions/48529625/in-pytorch-how-are-layer-weights-and-biases-initialized-by-default), both weights and bias are initialized in pytorch based on a uniform distribution with standard deviation inversely proportional to the square root of the total fanin (which is much bigger than the typical effective fanin).

**Comment by Sopel97** [comment]
We tried to match nodchip initialization early on, especially because the fact that the input layer is very sparse as you mentioned, but the results were worse. We have not tried to change initialization only for the feature transformer. I think we should do it and see how the nets look like then.

**Comment by ddobbelaere** [comment]
The discovered dead neurons provide an explanation for the lower compressed file size, so closing this.
